File: Level2/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    int current = 0;
    int save = 0;
    queue<int> going;
    
    for(int i = 0; i < bridge_length; i++)
        going.push(0);

    while(!going.empty())
    {
        answer++;
        going.pop();
        current -= save;
        if(!truck_weights.empty())
        {
            if(current + truck_weights.front() <= weight)
            {
                current += truck_weights.front();
                going.push(truck_weights.front());
                truck_weights.erase(truck_weights.begin());
            }
            else
                going.push(0);
        }
        save = going.front();
    }
    return answer;
}

File: Level2/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let answer = 0;
    let sum = 0;
    let queue = new Array(bridge_length);
    
    queue.fill(0);
    
    while(queue.length > 0){
        answer++;
        
        sum -= queue.shift();
        
        
        if(truck_weights.length > 0){
            if(sum + truck_weights[0] <= weight){
                let truck = truck_weights.shift();
                sum += truck;
                queue.push(truck);
            }
            else{
                queue.push(0);
            }
        }
    }
    return answer;
}